### PR TITLE
speeding up of app load: 40 seconds -> 4 seconds

### PR DESCRIPTION
**Pull request by knorth55**
UPDATE: I update this PR to speed up more.

Loading apps takes long time, which drops my productivity.
~I change not to use `roslib.packages.find_resource` for speed up.~
I change to share the same `rospack` object between multiple functions to make  `roslib.packages.find_resource` faster.

Current `kinetic-devel` takes 39 seconds to load 12 app.
This PR reduces it to 4 seconds.

## `kinetic-devel` branch

```
[ERROR] [1640440511.654021]: total time: 39.4751198292
```

cProfile output

```
         67248708 function calls (62972939 primitive calls) in 39.464 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   39.463   39.463 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app_list.py:215(update)
        1    0.000    0.000   39.462   39.462 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app_list.py:167(_load)
       14    0.000    0.000   39.451    2.818 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app_list.py:122(update)
        7    0.000    0.000   39.451    5.636 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app_list.py:93(__init__)
        7    0.000    0.000   39.451    5.636 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app_list.py:101(_load)
       25    0.000    0.000   39.431    1.577 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:306(load_AppDefinition_by_name)
      100    0.041    0.000   39.352    0.394 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:117(find_resource)
      100    0.002    0.000   39.310    0.393 /opt/ros/melodic/lib/python2.7/dist-packages/roslib/packages.py:487(find_resource)
       25    0.001    0.000   29.561    1.182 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:278(load_AppDefinition_from_file)
      100    0.027    0.000   27.902    0.279 /opt/ros/melodic/lib/python2.7/dist-packages/catkin/find_in_workspaces.py:95(find_in_workspaces)
    12148    0.062    0.000   27.395    0.002 /usr/local/lib/python2.7/dist-packages/catkin_pkg/packages.py:75(find_packages)
    12148    0.084    0.000   27.324    0.002 /usr/local/lib/python2.7/dist-packages/catkin_pkg/packages.py:112(find_packages_allowing_duplicates)
    12148    1.674    0.000   25.028    0.002 /usr/local/lib/python2.7/dist-packages/catkin_pkg/package.py:583(parse_package_string)
4152808/188764    4.308    0.000   11.479    0.000 /usr/lib/python2.7/copy.py:145(deepcopy)
      100    0.000    0.000   11.206    0.112 /usr/lib/python2.7/dist-packages/rospkg/rospack.py:199(get_path)
      100    0.117    0.001   11.205    0.112 /usr/lib/python2.7/dist-packages/rospkg/rospack.py:173(_update_location_cache)
    32500    0.863    0.000   11.088    0.000 /usr/lib/python2.7/dist-packages/rospkg/rospack.py:49(list_by_path)
   188764    1.031    0.000    9.937    0.000 /usr/lib/python2.7/copy.py:306(_reconstruct)
       25    0.000    0.000    9.837    0.393 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:191(_AppDefinition_load_interface_entry)
       25    0.000    0.000    9.836    0.393 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:153(_AppDefinition_load_icon_entry)
       25    0.000    0.000    9.832    0.393 /home/knorth55/ros/vive_ws/src/PR2/app_manager/src/app_manager/app.py:174(_AppDefinition_load_launch_entry)
   377528    1.017    0.000    7.454    0.000 /usr/lib/python2.7/copy.py:234(_deepcopy_tuple)
     6472    0.296    0.000    5.672    0.001 /usr/local/lib/python2.7/dist-packages/catkin_pkg/package.py:115(__getattr__)
327164/163962    0.888    0.000    5.613    0.000 /usr/lib/python2.7/os.py:209(walk)
```

## This PR

```
[ERROR] [1640619177.785652]: total time: 4.10533118248
```

### CODE REVIEW

**Comment by knorth55** [comment]
I update this PR, and now it only takes 1 seconds to load the app definition.

@708yamaguchi can you check this PR in your PC and report how fast it is?

**Review by 708yamaguchi — COMMENTED** [review]
I think this is very useful PR.

I have 619 ROS packages in my workspaces,
```
$ rospack list | wc -l
619
```
and I have 18 apps in 4 packages in my workspaces.
```
[INFO] [1640495997.563157]: 1 apps found in /home/leus/ros/melodic/src/jsk-ros-pkg/jsk_pr2eus/pr2eus_tutorials/apps/app.installed
[INFO] [1640495999.662914]: 2 apps found in /home/leus/ros/melodic/src/jsk-ros-pkg/jsk_robot/jsk_robot_common/jsk_robot_startup/apps/robot_apps.installed
[INFO] [1640496012.381850]: 12 apps found in /home/leus/ros/melodic/src/jsk-ros-pkg/jsk_robot/jsk_fetch_robot/jsk_fetch_startup/apps/fetch_apps.installed
[INFO] [1640496015.567683]: 3 apps found in /home/leus/ros/melodic/src/jsk-ros-pkg/jsk_robot/jsk_pr2_robot/jsk_pr2_startup/apps/pr2_apps.installed
```

Without this PR,
```
35818241 function calls (33742020 primitive calls) in 19.040 seconds
```

```
[ERROR] [1640496024.228704]: total time: 27.6996641159 
```

With this PR,
```
478007 function calls (462837 primitive calls) in 0.423 seconds
```

```
[ERROR] [1640495916.035495]: total time: 0.429724931717
```

File: src/app_manager/app.py [hunk 1]
```diff
@@ -153,12 +153,12 @@ def load_Interface_from_file(filename):
...
         y = yaml.load(f.read())
-        y = y or {} #coerce to dict
-        try:
-            subscribed_topics = y.get('subscribed_topics', {})
-            published_topics = y.get('published_topics', {})
-        except KeyError:
-            raise InvalidAppException("Malformed interface, missing keys")
+    y = y or {} #coerce to dict
+    try:
+        subscribed_topics = y['subscribed_topics'] if 'subscribed_topics' in y else {}
```
Comment (v4hn) [inline src/app_manager/app.py]: I don't see how converting all these places to a more verbose syntax helps with the speedup.
At least with my local profiling there is literally no difference between `a.get(b,'')` and `a[b] if b in a else ''`.
What's your benchmarking result compared to leaving all these replacements out?

File: src/app_manager/app.py [hunk 2]
```diff
@@ -126,7 +127,18 @@ def find_resource(resource):
...
+    # roslib.packages.find_resource is too slow
+    # matches = roslib.packages.find_resource(p, a)
+    if rospack is None:
+        rospack = rospkg.RosPack()
+    pkg_path = rospack.get_path(p)
+    matches = []
+    for dirpath, dirnames, filenames in os.walk(pkg_path):
+        for filename in filenames:
+            if filename == a:
+                matches.append(os.path.join(dirpath, filename))
```
Comment (v4hn) [inline src/app_manager/app.py]: Are you sure this works well for devel as well as install spaces, isolated and merged?
`find_resources` has a rather elaborate implementation for several layouts. That's likely for historical reasons as well as they mention "the new catkin build system", but probably not exclusively?

Also I just found `find_resources` supports an optional `rospack` parameter and explicitly caches in it! What happens if you don't change anything but make a global rospack instance to reuse here?

Comment (knorth55) [reply src/app_manager/app.py]: I didn't understand the `find_resources`.
I will change that.

File: src/app_manager/app.py (hunk 1 above)
Comment (knorth55) [reply src/app_manager/app.py]: In my env, it got faster a little bit less than 1 seconds.
But it is really a small difference.

File: src/app_manager/app.py (hunk 2 above)
Comment (knorth55) [reply src/app_manager/app.py]: I change to use `find_resource` with `rospack`, and it takes 4 seconds.

```
[ERROR] [1640619177.785652]: total time: 4.10533118248
```

File: src/app_manager/app.py (hunk 1 above)
Comment (knorth55) [reply src/app_manager/app.py]: with `dict.get`

```
[ERROR] [1640619851.553788]: total time: 4.12315011024
```

no `dict.get`

```
[ERROR] [1640619765.145302]: total time: 3.93386387825
```

Comment (knorth55) [reply src/app_manager/app.py]: However, `dict.get` makes no big change, so I will remove the change.

Comment (v4hn) [reply src/app_manager/app.py]: part of that might be noise, I wouldn't even expect 0.2s difference based on my minor tests here. But you are right, even 0.2s is not enough to justify these changes :+1:

**Review by v4hn — APPROVED** [review]
I didn't test any of this but now the changes make sense looking at the source code.
Please review and merge @k-okada .

**Comment by Affonso-Gui** [comment]
@k-okada Kindly ping

**Pull request merged**